### PR TITLE
Eliminate compiler warnings when building HIP code for AMD

### PR DESCRIPTION
**Pull request by jakurzak**
## Proposed changes

A few simple fixes to eliminate all lingering compiler warnings when building for AMD GPUs.

## What type(s) of changes does this code introduce?

elimination of warnings

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

- AMD Radeon VII
- NVIDIA A100

## Checklist

- Yes, this PR is up to date with the current state of 'develop'
- Clang formatting N/A
- Adding tests N/A
- Documentation N/A

### CODE REVIEW

File: src/Particle/Lattice/ParticleBConds3DSoa.h [hunk 1]
```diff
@@ -428,7 +428,7 @@ struct DTD_BConds<T, 3, PPPG + SOA_OFFSET>
 
       T rmin = delx * delx + dely * dely + delz * delz;
       int ic = 0;
-#pragma unroll(7)
```
Comment (ye-luo) [inline src/Particle/Lattice/ParticleBConds3DSoa.h:431]: This seems to be a syntactic change. Does this really change anything?

**Review by prckent — APPROVED** [review]
LGTM

Thanks for the cleanup.

**Comment by prckent** [comment]
Test this please

**Review by ye-luo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -798,7 +798,7 @@ if(ENABLE_ROCM)
   find_package(rocthrust CONFIG REQUIRED)
   # architecture flags
   set(HIP_ARCH
-      "gfx906,gfx908"
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply CMakeLists.txt:801]: No objections.
Will change to gfx906 only.

File: src/Particle/Lattice/ParticleBConds3DSoa.h (hunk 1 above)
Comment (jakurzak) [reply src/Particle/Lattice/ParticleBConds3DSoa.h:431]: IMO, it's questinable at best to tell the compiler how much to unroll.
But that's that the code does. Not my call.
I've never seen the number in parentheses, tough, and hipcc complains about it.
I am leaning towards leaving it without parentheses.

Comment (ye-luo) [reply src/Particle/Lattice/ParticleBConds3DSoa.h:431]: Sorry I didn't express my question clearly. I'm not asking about how good unroll works. I'm wondering if changing syntax really has impact.
I saw more compiler support with parentheses being added.
Clang clearly supports it
https://clang.llvm.org/docs/AttributeReference.html#pragma-unroll-pragma-nounroll

This file should not be compiled by hipcc. Do you have the warning at hand?

Comment (jakurzak) [reply src/Particle/Lattice/ParticleBConds3DSoa.h:431]: Good point.
It's meant for the host compiler.
It's only a warning with hipcc, which I can easily suppress with a flag.
Let me undo this then.

**Comment by jakurzak** [comment]
I think I fixed all objectionable changes.
I really appreciate your thorough review and your patience.

**Review by ye-luo — DISMISSED** [review]

**Comment by ye-luo** [comment]
Test this please

File: src/Particle/Lattice/ParticleBConds3DSoa.h (hunk 1 above)
Comment (jakurzak) [reply src/Particle/Lattice/ParticleBConds3DSoa.h]: BTW, the warning is, e.g.:
/mnt/shared/container-share/qmcpack/src/Particle/Lattice/ParticleBConds3DSoa.h:431:16: warning: argument to '#pragma unroll' should not be in parentheses in CUDA C/C++ [-Wcuda-compat]
#pragma unroll(7)

Comment (ye-luo) [reply src/Particle/Lattice/ParticleBConds3DSoa.h]: Likely you were using hipcc as your host compiler. We don't recommend this way.

**Review by prckent — APPROVED** [review]

**Pull request merged**